### PR TITLE
[SILVerifier] Verify that the line table is contiguous at SIL level.

### PR DESCRIPTION
**Pull request by dcci**
<rdar://problem/44317688>

I don't plan to merge this until I sorted out large part of the verification failures this introduced [and after I'll introduce a way to disable this], but the general idea should be there [happy to be proven wrong]. Decoding debug location turns out it might be expensive, but I think it's too early to make measurements.

### CODE REVIEW

File: lib/SIL/SILVerifier.cpp [hunk 1]
```diff
@@ -4926,6 +4968,7 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
     
     SILInstructionVisitor::visitSILBasicBlock(BB);
     verifyDebugScopeHoles(BB);
+    verifyLineTable(BB);
```
Comment (adrian-prantl) [inline lib/SIL/SILVerifier.cpp]: Is there a mechanism for "expensive" verifiers?

**Comment by dcci** [comment]
@atrick This is now ready for review.

File: lib/SIL/SILVerifier.cpp [hunk 2]
```diff
@@ -4838,6 +4838,42 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+        Init = true;
+        continue;
+      }
+      if (DL.Line >= LastLine || LastFilename != DL.Filename) {
+        LastLine = DL.Line;
+        LastFilename = DL.Filename;
+        continue;
+      }
+      require(DL.Line >= LastLine, "Line table is not monotonically increasing"
+                                   " within a single basic block");
```
Comment (jrose-apple) [inline lib/SIL/SILVerifier.cpp]: I don't think this is a correct assertion anyway. Consider

```
let x = foo() +
  bar() +
  baz()
```

File: lib/SIL/SILVerifier.cpp (hunk 1 above)
Comment (jrose-apple) [reply lib/SIL/SILVerifier.cpp]: We did one for the ASTVerifier by checking `SWIFT_DISABLE_AST_VERIFIER` in addition to `NDEBUG`. You could add something like that too, or put it behind an -Xllvm flag (which IIRC won't be available in Release builds at all).

Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: You're right. Back to the whiteboard ..

**Review by atrick — COMMENTED** [review]
This makes sense to me. I guess you can commit as soon as all the tests are passing, including SCK.

**Review by adrian-prantl — CHANGES_REQUESTED** [review]
This is a great start!

File: lib/SIL/SILVerifier.cpp [hunk 3]
```diff
@@ -62,6 +62,10 @@ static llvm::cl::opt<bool> AbortOnFailure(
                               "verify-abort-on-failure",
                               llvm::cl::init(true));
 
+static llvm::cl::opt<bool> VerifyLineTable(
+                              "verify-line-table",
```
[the remark this answers is not in the document]
Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: Done.

File: lib/SIL/SILVerifier.cpp [hunk 4]
```diff
@@ -4838,6 +4842,56 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
     }
   }
 
+  void verifyLineTable(SILBasicBlock *BB) {
+    if (!VerifyLineTable)
+      return;
+
+    SILFunction *F = BB->getParent();
+    if (F->getEffectiveOptimizationMode() != OptimizationMode::NoOptimization)
```
[the remark this answers is not in the document]
Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: Done

File: lib/SIL/SILVerifier.cpp [hunk 5]
```diff
@@ -4838,6 +4842,56 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+      return;
+
+    if (F->isTransparent())
+      return;
+
+    auto &SrcMgr = BB->getParent()->getModule().getSourceManager();
+    uint64_t LastLine = 0;
```
[the remark this answers is not in the document]
Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: Gone away.

File: lib/SIL/SILVerifier.cpp [hunk 6]
```diff
@@ -4838,6 +4842,56 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+    using UUSTuple = std::pair<std::pair<unsigned, unsigned>, StringRef>;
+    struct DebugLocKey : public UUSTuple {
+      DebugLocKey(SILLocation::DebugLoc DL)
+          : UUSTuple({{DL.Line, DL.Column}, DL.Filename}) {}
+      inline bool operator==(const SILLocation::DebugLoc &DL) const {
+        return first.first == DL.Line && first.second == DL.Column &&
+               second.equals(DL.Filename);
+      }
+    };
+    llvm::DenseSet<UUSTuple> VisitedLocations;
```
[the remark this answers is not in the document]
Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: How we can detect line zero?

File: lib/SIL/SILVerifier.cpp [hunk 7]
```diff
@@ -4838,6 +4842,56 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+          std::tie(DL.Line, DL.Column, DL.Filename))
+        continue;
+
+      auto ItNew = DebugLocKey(DL);
+
+      require(!VisitedLocations.count(ItNew), "Incorrect location at -Onone");
+      VisitedLocations.insert(ItNew);
+      LastLine = DL.Line;
+      LastColumn = DL.Column;
+      LastFilename = DL.Filename;
```
[the remark this answers is not in the document]
Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: Done, thanks.

File: lib/SIL/SILVerifier.cpp (hunk 1 above)
Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: Now the algorithm should work.

File: lib/SIL/SILVerifier.cpp [hunk 8]
```diff
@@ -4847,6 +4851,54 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+      if (isa<MetatypeInst>(&SI))
+        continue;
+      auto LocKind = SI.getDebugLocation().getLocation().getKind();
+      if (LocKind != SILLocation::LocationKind::RegularKind)
+        continue;
+      auto DL = SI.getDebugLocation().getLocation().decodeDebugLoc(SrcMgr);
+      if (DL.Line == 0)
+        continue;
+      auto ItNew = DebugLocKey(DL);
+      require(!VisitedLocations.count(ItNew), "Incorrect location at -Onone");
```
Comment (jrose-apple) [inline lib/SIL/SILVerifier.cpp]: I don't think we make any promise that two different AST expressions can't have the same location, and we certainly don't promise that any particular AST expression only corresponds to one SIL instruction. What are you actually trying to test?

Comment (dcci) [reply lib/SIL/SILVerifier.cpp]: The idea we came up with Adrian is that of trying to add a pass to make sure we don't have locations jumping back within a single basic block.
The motivation is that, even at `-Onone`, we see sometimes stepping in the debugger jumping back in straight line code, and we thought we could add a verifier similarly to what we did for debug scopes [which helped catching a lot of bugs].

To give you some more context, the verifier was originally written in `IRGenDebugInfo.cpp`, and there are still some vestiges of it there (`-verify-linetable`), then we decided to move it at SIL level, so we can catch which optimization breaks the invariants running `-sil-verify-all`.

The main invariant that the current code tries to maintain is making sure that we don't end up bouncing back to an existing location, but from what you say this is something we shouldn't rely on? Do you think there's a better way of achieving what we're trying to do here?

Thanks!

File: lib/SIL/SILVerifier.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrian-prantl) [reply lib/SIL/SILVerifier.cpp]: You probably noticed how SILLocation wraps three different storage mechanisms. A line 0 location will always be backed by an already-decoded location, so we could add a very cheap isCompilerGenerated() method to SILLocation that detects it by first checking the storage kind and then, if applicable, for line number 0.

File: lib/SIL/SILVerifier.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrian-prantl) [reply lib/SIL/SILVerifier.cpp]: You probably noticed how SILLocation wraps three different storage mechanisms. A line 0 location will always be backed by an already-decoded location, so we could add a very cheap isCompilerGenerated() method to SILLocation that detects it by first checking the storage kind and then, if applicable, for line number 0.

File: lib/SIL/SILVerifier.cpp (hunk 8 above)
Comment (jrose-apple) [reply lib/SIL/SILVerifier.cpp]: I think it'd be reasonable to say that you'd never go backwards to a location in a previous *statement,* as well as making a policy of not revisiting the same AST *expression.* But I don't think you can do the latter based on source locations, and the former requires using a ParentMap or something to know which locations are within which statements.

Comment (jrose-apple) [reply lib/SIL/SILVerifier.cpp]: Even that gets tricky with inlined closures.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**